### ISSUE TITLE
[Meta] Moving Staking off the Relay Chain

### ISSUE THREAD

**Issue by Ank4n** [paritytech/polkadot#7657]
Some context on why we should do this => https://github.com/paritytech/polkadot/issues/323.

We had an in-depth discussion about design of a Staking Parachain in the last Frame Retreat in July with Gavin also present. This initial writeup of the issue is mostly inspired from those discussion.

### pallet-balances

DOT tokens can be transferred to staking chain in order to stake. The chain cannot burn or mint the tokens. Relay chain (in future this could be Asset hub) could mint 8% of total supply and teleport it to the Staking Chain at the start of each cycle where each cycle is 28 eras. This 8% then is only used for Staking rewards.

We talked about not having tokens on Staking chain at all (or may be a small balance just for paying fees) but eventually decided against the complexity introduced because of it. The topic is a bit beyond the scope of this issue, though we could revisit it if someone has strong opinions about it.

### Staking logic

Most of the staking logic can be reused from the Relay chain but there might be need for significant refactorings to make the staking related pallets PoV friendly.

### Parallel Staking on Relay Chain and Staking Chain

We want to scale the staking parachain gradually. This is an extremely important piece of logic that needs to be tested thoroughly, but even in production, we should try taking all the steps to mitigate against potential bugs/oversights.

I will refer staking module to mean set of pallets on relay chain that contains the staking and validator election logic. The staking module interacts with the rest of relay chain in two ways:

- pallet-session: Asks Staking module for a new set of validators (and their exposure) at start of each new session.
- pallet-offences: Reports offences by validators to Staking module.

The desired set of validator count is set today on the staking module. We can introduce another config to determine part of the validators to get from the staking parachain, and the rest we will elect from the local staking module itself as it happens today. We can start with a smaller number (10%) and eventually scale this to 100% from Staking parachain.

The important thing to note is staking would work independently on both staking chain and relay chain. An account who wants to validate on staking chain would have to transfer DOTs to staking chain and register their intention to validate there. Same for nominators.

The disadvantage of this approach is, all existing validators (and nominators) will have to start over and they will lose their current set of nominators (and nominations).

### Unknowns/Needs further investigation

- Validators need to set session keys on the session pallet and tie it to their ValidatorID (stash account). The validators on staking chain would probably need to set session keys on the relay chain with a remote origin from Staking chain. Would this work?

- While we are in transition period, if staking parachain fails to perform a new election, we can configure staking module to elect more validators locally as backup. Once staking parachain is 100% live though, we should stop allowing new validators on relay chain (potentially even kicking them out). What would happen if Staking parachain is bricked or for some reason not able to perform new elections? Should we have some kind of back up on relay chain? What kind of backup would we need?

- I don't understand slashing logic too well and am not 100% sure how would slashing work with Staking Parachain. Currently, it seems pallet session gets the exposure information as well for the session validator set, generates a merkle trie and stores the root of the trie for each session. Can this be done somehow without sending exposure data? Do we need XCQ or similar before we can do this?

#### XCM primitives + Runtime Apis

Moving staking logic to a parachain might need significant refactoring to current pallets since these were not designed to be POV friendly. We could also take this chance to fork the existing pallets, do some major refactorings or even rewrite some pallets from scratch. @kianenigma had some ideas on [refactoring staking pallet into smaller pallets](https://forum.parity.io/t/mega-refactor-repot-staking-into-smaller-pallets/754) and also of [pagifying everything](https://youtu.be/qVd9lAudynY?t=3735) (elections, rewards, slashing) in staking. We can also clean up controller/stash logic from staking and use proxy instead.

This might mean the current staking pallet and the new staking parachain pallets might diverge a lot in their interface (dispatchable functions and storage items). Wallets/dapps/parachains that interact with staking pallet today might need to be updated to work with the staking parachain. We could start with creating a set of XCM primitives and runtime apis for staking, giving time for builders to start adopting to that before Staking Parachain goes live. The staking parachain could adhere to same XCM primitives and runtime apis making switching to staking parachain would be easier. This would be especially useful while we are in transition period and staking parachain is not providing all the validators.

### Next steps
- Get feedback on this issue.
- [Map out current implementation](https://hackmd.io/Ljs1638oSceFQnmPaq4GSQ) of staking related pallets and the work to refactor them (or rewrite them).
- Make a proper plan of action to execute this.

**Comment by burdges** [comment]
Do we know how much relay chain usage is staking related now?

**Comment by joepetrowski** [comment]
> We talked about not having tokens on Staking chain at all (or may be a small balance just for paying fees) but eventually decided against the complexity introduced because of it. The topic is a bit beyond the scope of this issue, though we could revisit it if someone has strong opinions about it.

This is more XCM v4 related, but it'd still be nice to have `Stake` and `Unstake` instructions in v4. A value proposition of Asset Hub is that it gives custodians the ability to support parachain native assets without running infrastructure for those parachains. But a lot of their users also want to stake, and if they have to teleport the assets back to the parachain to make certain calls, then it puts them back in the same problem of needing to run nodes there.

So, we don't have to use these instructions for DOT staking, but it'd be nice to have them in XCM v4 so that other chains can impl those instructions, and on Asset Hub we can have a `stake_remote(asset, amount)` call.
